### ISSUE TITLE
[Discussion] Vision for daskernetes

### ISSUE THREAD

**Issue by jacobtomlinson** [pangeo-data/pangeo#89]
We've just had a long discussion within our team about what we are aiming for with integrating dask into our jupyterhub setup. This is going to be a bit of a brain dump and I would like feedback from @mrocklin and @yuvipanda especially.

There was talk yesterday about whether we need daskernetes and the help chart to both exist. I think there may be some confusion as to what we want the helm chart to do, and from our side what daskernetes is trying to achieve.

For a while now I have envisaged something similar to JupyterHub existing for dask, let's call it DaskHub for now. This would be a separate service which can create dask schedulers and workers via a set of spawners. I would imagine that it has an API and so it could be integrated nicely with JupyterHub and notebooks/lab. But it could also be a separate thing which could be deployed independently of JupyterHub if required. My aspiration for a dask helm chart would be a chart that deploys DaskHub, not a chart which deploys a single cluster (which is what currently exists in kubernetes/stable).

My vision for pangeo (and Jade previously) would be a higher level helm chart which has a dependancy on both the JupyterHub chart and the DaskHub chart, these would be configured for JupyterHub to operate the DaskHub and the user would not need to interact with DaskHub directly. Basically pangeo would be a curated set of helm charts which make up a complete service with all the configuration hard work figured out for some specific platforms (GCE, AWS, Azure, etc).

I also see pangeo introducing some standards for user workspaces which will ensure they will run on pangeo. For example if your organisations workspace docker image has dependency x, y and z at versions k, l and m or higher then it can be dropped into pangeo and "just work" as the image for the notebook/lab, the scheduler and the worker.

From following along with daskernetes and engaging with you folks yesterday makes me think that daskernetes is not DaskHub. My understanding is that daskernetes is a tool that a user would use to create their own scheduler and workers, basically a spawner that is operated manually.

I guess initial feedback I would like is:

- Have I understood what daskernetes is correctly?
- Do you have thoughts on what I propose DaskHub could be?
- Or anything else 😄 

cc @niallrobinson

**Comment by mrocklin** [comment]
-  You are correct about the current daskernetes library.  It allows someone on a kubernetes cluster to launch their own standalone dask cluster with a local scheduler and a set of remote workers.
-  I agree that something like DaskHub would be useful.  I see the following benefits
    -  A shared place to enforce user resource limits in ways that are perhaps more complex than default Kubernetes policies
    - A central place for logging and diagnostics

There are a few dask-on-kubernetes things that exist or have been proposed

-  Daskernetes: dynamic control of worker pods from a kubernetes pod
-  Dask Helm chart: static deployment of 1 scheduler and n workers in a simple and non-responsive way
-  DaskHub: central control of many dask clusters

I want to be clear that none of these are "the way" to deploy dask on Kubernetes.  It's more likely that all of them will exist.  What we need for Pangeo and what others will need for their applications like differ substantially.  In regards to the helm charts in particular I think that making a dask helm chart now does not stop us from making a daskhub helm chart later.  The proposed DaskHub idea is opinionated enough that I would not want it to carry the name of "the dask chart", much as JupyterHub does not carry the name Jupyter.

For Jade/Pangeo in particular can you state some advantages of the DaskHub (centrally controlled) over the Daskernetes (user controlled) solution?  Are you mostly looking for central management of user resources?

**Comment by jacobtomlinson** [comment]
I agree that none of these things should be "the way" in the same way JupyterHub is not "the way" of deploying multi-user notebook setups. 

We are striving for an environment where users are getting a certain amount of functionality out of the box and behind the scenes. Then if they wish to gain a stronger understanding of what is going on they can make their workflows even more powerful.

The example I'm thinking of is an Iris based workflow. We have a branch of Iris which now uses dask under the hood instead of Biggus. This means that a user who already knows how to use Iris can log onto our JupyterHub setup and when they use Iris it will parallelise seamlessly on the dask scheduler. This gives them **some** performance boost "for free" in terms of their code. They can also be more intelligent about how they are writing their code and with dask in mind, which will give them even more performance increases. 

The idea behind this is to slowly change their mindset to a more parallel one by making parallelisation the path of least resistance. If we provided them with a notebook and then told them they had to use daskernetes to create a cluster there is already a learning curve and mental hurdle to navigate.

If we provide them a scheduler with a small number of workers out of the box they will just get a better experience to working locally. Then when they are more comfortable we can start introducing them to the fact they can refactor their code and change the size of the cluster to get an even better experience.

**Comment by mrocklin** [comment]
I think that we could probably auto-launch a dask cluster for them with any of these approaches
